synapse/tests/test_lib_scrape.py

```diff
@@ -704,3 +704,122 @@ def test_scrape_context(self):
+        cpe:2.3:*:langtest:*:*:*:*:-:*:*:*:*
+        cpe:2.3:*:langtest:*:*:*:*:*:*:*:*:*
+        cpe:2.3:*:langtest:*:*:*:*:en:*:*:*:*
+        cpe:2.3:*:langtest:*:*:*:*:usa:*:*:*:*
+        cpe:2.3:*:langtest:*:*:*:*:usa-en:*:*:*:*
+        cpe:2.3:*:langtest:*:*:*:*:usa-123:*:*:*:*
+
+        A few examples
+        cpe:2.3:a:ntp:ntp:4.2.8:p3:*:*:*:*:*:*
+        cpe:2.3:o:microsoft:windows_7:-:sp2:*:*:*:*:*:*
+        cpe:2.3:a:hp:insight:7.4.0.1570:-:*:*:online:win2003:x64:*
+        cpe:2.3:a:foo\\bar:big\$money_2010:*:*:*:*:special:ipod_touch:80gb:*
+        cpe:2.3:a:hp:openview_network_manager:7.51:*:*:*:*:linux:*:*
+        cpe:2.3:a:apple:swiftnio_http\/2:1.19.1:*:*:*:*:swift:*:*
+
+        Some quoted examples
+        cpe:2.3:a:fooo:bar_baz\:_beep_bpp_sys:1.1:*:*:*:*:ios:*:*
+        cpe:2.3:a:lemonldap-ng:apache\:\:session\:\:browsable:0.9:*:*:*:*:perl:*:*
+        cpe:2.3:a:daemon-ng:hurray\:\::0.x:*:*:*:*:*:*:*
+        cpe:2.3:a:microsoft:intern\^et_explorer:8.0.6001:beta:*:*:*:*:*:*
+
+        TEXT examples
+        Example double quoted cpe value "cpe:2.3:a:vertex:synapse:*:*:*:*:*:*:*:dquotes".
+        Example double quoted cpe value "cpe:2.3:a:vertex:synapse:*:*:*:*:*:*:*:MiXeDcAsE".
+        Example single quoted cpe value "cpe:2.3:a:vertex:synapse:*:*:*:*:*:*:*:squotes".
+        A CPE at the end of a sentence like this captures the period... cpe:2.3:a:*:*:*:*:*:*:*:*:*:hasperiod.
+        Some CPE are exciting! Like this cpe:2.3:a:*:*:*:*:*:*:*:*:*:noexclaim!
+        Some CPE are boring! Like this cpe:2.3:a:*:*:*:*:*:*:*:*:*:noslash\
+        Unicode endings are omitted cpe:2.3:a:*:*:*:*:*:*:*:*:*:unicodeend0ॐ
+        Unicode quotes “cpe:2.3:a:*:*:*:*:*:*:*:*:*:smartquotes”
+
+        EMBEDDED TEXT
+        wordscpe:2.3:a:vendor:product:version:update:edition:lng:sw_edition:target_sw:target_hw:otherxxx:newp
+        wordscpe:2.3:a:vendor:product:version:update:edition:lng:sw_edition:target_sw:target_hw:otherzzz:
+
+        BAD values
+        cpe:2.3:a:vertex:synapse:*:*:*:NEWP:*:*:*:*
+        cpe:2.3:a::::::::::
+        cpe:2.3:a:vendor:product:version:update:edition:lng:sw_edition:target_sw:ॐ:other
+        cpe:2.3:a:vendor:product:version:update:edition
+        cpe:2.3:a:opps:bad_quote\\/2:1.19.1:*:*:*:*:swift:*:*
+
+        # Bad languages
+        cpe:2.3:*:langtest:*:*:*:*:a:*:*:*:*
+        cpe:2.3:*:langtest:*:*:*:*:aaaa:*:*:*:*
+        cpe:2.3:*:langtest:*:*:*:*:usa-o:*:*:*:*
+        cpe:2.3:*:langtest:*:*:*:*:usa-omn:*:*:*:*
+        cpe:2.3:*:langtest:*:*:*:*:usa-12:*:*:*:*
+        cpe:2.3:*:langtest:*:*:*:*:usa-1234:*:*:*:*
+        '''
+        nodes = sorted(set(s_scrape.scrape(cpedata, ptype='it:sec:cpe')))
+        nodes.remove(('it:sec:cpe', 'cpe:2.3:*:*:*:*:*:*:*:*:*:*:*'))
+        nodes.remove(('it:sec:cpe', 'cpe:2.3:*:-:na:*:*:*:*:*:*:*:*'))
+        nodes.remove(('it:sec:cpe', 'cpe:2.3:*:.:dot:*:*:*:*:*:*:*:*'))
+        nodes.remove(('it:sec:cpe', 'cpe:2.3:*:?:spec1:*:*:*:*:*:*:*:*'))
```

Comment:
is this one actually valid? it doesn't look like a single unquoted ? is a valid avstring

Reply:
It's not valid :)